src/main/java/org/alfresco/repo/domain/schema/SchemaBootstrap.java

```diff
@@ -884,22 +902,41 @@ private boolean updateSchema(Connection connection) throws Exception
 
             if (logger.isInfoEnabled())
             {
-                logger.info("Create scripts executed in "+(System.currentTimeMillis()-start)+" ms");
+                logger.info("Create scripts executed in " + (System.currentTimeMillis() - start) + " ms");
             }
         }
         else
         {
+            long start = System.currentTimeMillis();
+
             // Execute any pre-auto-update scripts
             checkSchemaPatchScripts(connection, preUpdateScriptPatches, true);
             
             // Execute any post-auto-update scripts
             checkSchemaPatchScripts(connection, postUpdateScriptPatches, true);
+
+            if (logger.isInfoEnabled())
+            {
+                logger.info(
+                    "Update case: checkSchemaPatchScripts scripts executed in " + (System.currentTimeMillis() - start) + " ms");
+            }
         }
 
+        ensureCurrentClusterMemberIsBootstrapping(connection);
+
         // Initialise Activiti DB, using an unclosable connection
         boolean activitiTablesExist = checkActivitiTablesExist(connection);
+        if (logger.isInfoEnabled())
+        {
+            logger.info("Activiti tables exist: " + activitiTablesExist);
+        }
+
         if(!activitiTablesExist)
         {
+            long start = System.currentTimeMillis();
+
+            ensureCurrentClusterMemberIsBootstrapping(connection);

```

Comment:
Why can't we check this only once for the whole bootstrap?

Reply:
We could, but that would mean important refactoring of the SchemaBootstrap class. The code was design (as see in this comment from 2008:"// Retries are required here as the DB lock will be applied lazily upon first statement execution.") to back off only if changes are required to the db AND can't create the db lock maker(alf_bootstrap_lock).

How can you know in a node if changes are required?
Only by going through the initiation steps:
1. check if the ALF tables need to be created- and create them
2. check if the ALF tables need to be patched - and apply them
3. check if the Activiti tables are present - and create them
4. check if the Activiti tables need patches -and apply them

At what point can you determine if another node in the cluster is present? And tries to do the same thing? 
There is no hazelcast yet or any of the fancy DB lock mechanisms;

The way the code (with this additional fix) works is:  
1. all the nodes in the cluster can't wait around and assume that there is another node in the cluster that is doing the create/upgrade schema, so they, themselves, try to see if tables need to be created or updated
2. the *first node in the cluster that decides* it needs to do some changes in the DB, creates (if it can) that marker table (alf_bootstrap_lock) and marks in the object (executedStatementsThreadLocal) that it plans to execute the initialization/upgrade of the DB
3. all other nodes in the cluster, before executing ANY statement have to check if there is a marker in the DB telling them that some other node is doing the upgrade/initialization
4. Only one node should succeed in getting that alf_bootstrap_lock created, therefore only one node will do the initialization/upgrade
5. All other nodes will back off ONLY when they detect that changes are required; If changes are not required they will boot as fast/efficient as possible- not even caring if there are other nodes in the cluster at this point

Case "3. check if the Activiti tables are present" from above was not covered by these checks before this fix;
The problem with the Activiti table initialization (initialiseActivitiDBSchema) is that the code that actually creates the tables is in the Activiti code, and also the initialization of the objects required to create the Activiti tables takes a long time, and this helps with creating problems in the cluster synchronization(checks like checkActivitiTablesExist(connection); ) we have in the code;

Lets take the Activiti tables initialization ("3. check if the Activiti tables are present") example (2 nodes cluster for simplicity) BEFORE THIS FIX:
1. Both nodes boot up around the same time
2. First node was a little bit faster and finishes the ALF tables initialization (reaches line " boolean activitiTablesExist = checkActivitiTablesExist(connection);"); This means it was able to create the alf_bootstrap_lock marker table and also has some data in the executedStatementsThreadLocal variable;
3. Second node checks if the ALF tables are present- they are! and if they need upgrading - they don't! ; So the second node has not executed any statements in the DB up to now(executedStatementsThreadLocal should be empty);
4. Then both nodes check if the Activiti tables exist - they don't
5. The first node is starting initializing the Activiti beans and at some point creates the tables in the DB
6. (BEFORE THE FIX) the second node was doing the same thing, because there was no check or code to try to mark that this second node was trying to initialize the Activiti tables. It did not check if there is another bootstrap in progress before creating the Activiti beans that initialize the tables;
7, Obviously the second node would fail too boot because of these errors

Do you see any other sort of flagging/locking that would ensure that only one of the cluster node would do the DB initialization/upgrade without knowing IF there is even any other cluster member out there ?

Reply:
It looks like `alf_bootstrap_lock` table should exist for the whole time of the bootstrap (Alfresco and Activiti tables). The Activiti tables are always created after the Alfresco tables and that should be done by the same node. I don't think that the check for Activiti tables to exist if we just created Alfresco tables makes any sense.
I'm not sure if it is easy to refactor the code in this way. The current solution should work as well.